File: leetcode/4.py
# -*- coding: utf-8 -*-
# @Author: gunjianpan
# @Date:   2020-03-25 21:44:54
# @Last Modified by:   gunjianpan
# @Last Modified time: 2020-03-25 22:30:29

"""
4. Median of Two Sorted Arrays Hard

There are two sorted arrays nums1 and nums2 of size m and n respectively.

Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

You may assume nums1 and nums2 cannot be both empty.

## Example 1:
nums1 = [1, 3]
nums2 = [2]

The median is 2.0

## Example 2:
nums1 = [1, 2]
nums2 = [3, 4]

The median is (2 + 3)/2 = 2.5
Accepted 616,039 Submissions 2,157,660
"""


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        N, M = len(nums1), len(nums2)
        if N > M:
            N, M = M, N
            nums1, nums2 = nums2, nums1
        K = (N + M + 1) // 2
        left, right = 0, N
        while left < right:
            # print(left, right)
            t1 = left + (right - left) // 2
            t2 = K - t1
            if nums1[t1] > nums2[t2 - 1]:
                right = t1
            elif nums1[t1] == nums2[t2 - 1]:
                left = t1 + 1
                break
            else:
                left = t1 + 1
        t1 = left
        t2 = K - t1
        c1 = max(
            nums1[t1 - 1] if t1 else float("-inf"),
            nums2[t2 - 1] if t2 else float("-inf"),
        )
        if (N + M) % 2 == 1:
            return c1
        c2 = min(
            nums1[t1] if t1 < N else float("inf"), nums2[t2] if t2 < M else float("inf")
        )
        # print(t1, t2)
        return (c1 + c2) / 2

File: leetcode/4.java
class Solution {
  public double findMedianSortedArrays(int[] nums1, int[] nums2) {
    // O(log(m + n))。
    //二分，每次在2个数组中去掉 中位值/2个数
    int m = nums1.length;
    int n = nums2.length;

    int left = (n + m + 1) / 2;
    int right = (n + m + 2) / 2;
    double mid1 = 0;
    double mid2 = 0;
    int s1 = 0;
    int s2 = 0;
    int e1 = m - 1;
    int e2 = n - 1;
    mid1 = findmid(nums1, nums2, left, s1, s2, e1, e2);
    mid2 = findmid(nums1, nums2, right, s1, s2, e1, e2);
    System.out.println(left + " " + right);
    System.out.println(mid1 + " " + mid2);

    return (mid1 + mid2) / 2;
  }

  private double findmid(int[] nums1, int[] nums2, int k, int s1, int s2,
                         int e1, int e2) {
    System.out.println(k + " " + s1 + " " + s2);
    double mid = 0;
    int len1 = e1 - s1 + 1;
    int len2 = e2 - s2 + 1;
    //让 len1 的长度小于 len2，这样就能保证如果有数组空了，一定是 len1
    if (len1 > len2)
      return findmid(nums2, nums1, k, s2, s1, e2, e1);
    if (len1 == 0) {
      System.out.println("==" + s2 + " " + k);
      return nums2[s2 + k - 1];
    }

    int a = k / 2;
    int a1 = s1 + a - 1;
    int a2 = s2 + a - 1;
    if (a1 > e1)
      a1 = e1; //中间位置
    if (a2 > e2)
      a2 = e2; //中间位置

    if (k == 1) {
      mid = nums1[s1] < nums2[s2] ? nums1[s1] : nums2[s2];
    } else {

      if (nums1[a1] <= nums2[a2]) {
        mid = findmid(nums1, nums2, k - (a2 - s2 + 1), a1 + 1, s2, e1, e2);
        System.out.println(" " + a1 + " " + a2);
      }

      else {
        mid = findmid(nums1, nums2, k - (a1 - s1 + 1), s1, a2 + 1, e1, e2);
        System.out.println(" " + a1 + " " + a2);
      }
    }

    return mid;
  }
}

